subsys/testsuite/ztest/src/ztest_new.c

```diff
@@ -0,0 +1,619 @@
+	 * Because we reuse the same k_thread structure this would
+	 * causes some problems.
+	 */
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		k_thread_abort(&ztest_thread);
+	}
+#endif
+
+	if (!ret && mock_status == 1) {
+		PRINT("Test %s failed: Unused mock parameter values\n",
+		      test->name);
+		ret = TC_FAIL;
+	} else if (!ret && mock_status == 2) {
+		PRINT("Test %s failed: Unused mock return values\n",
+		      test->name);
+		ret = TC_FAIL;
+	} else {
+		;
+	}
+
+	return ret;
+}
+
+#ifdef KERNEL
+#ifdef CONFIG_SMP
+#define NUM_CPUHOLD (CONFIG_MP_NUM_CPUS - 1)
+#else
+#define NUM_CPUHOLD 0
+#endif
+#define CPUHOLD_STACK_SZ (512 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+static struct k_thread cpuhold_threads[NUM_CPUHOLD];
+K_KERNEL_STACK_ARRAY_DEFINE(cpuhold_stacks, NUM_CPUHOLD, CPUHOLD_STACK_SZ);
+static struct k_sem cpuhold_sem;
+volatile int cpuhold_active;
+
+/* "Holds" a CPU for use with the "1cpu" test cases.  Note that we
+ * can't use tools like the cpumask feature because we have tests that
+ * may need to control that configuration themselves.  We do this at
+ * the lowest level, but locking interrupts directly and spinning.
+ */
+static void cpu_hold(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	unsigned int key = arch_irq_lock();
+	uint32_t dt, start_ms = k_uptime_get_32();
+
+	k_sem_give(&cpuhold_sem);
+
+#if defined(CONFIG_ARM64) && defined(CONFIG_FPU_SHARING)
+	/*
+	 * We'll be spinning with IRQs disabled. The flush-your-FPU request
+	 * IPI will never be serviced during that time. Therefore we flush
+	 * the FPU preemptively here to prevent any other CPU waiting after
+	 * this CPU forever and deadlock the system.
+	 */
+	extern void z_arm64_flush_local_fpu(void);
+	z_arm64_flush_local_fpu();
+#endif
+
+	while (cpuhold_active) {
+		k_busy_wait(1000);
+	}
+
+	/* Holding the CPU via spinning is expensive, and abusing this
+	 * for long-running test cases tends to overload the CI system
+	 * (qemu runs separate CPUs in different threads, but the CI
+	 * logic views it as one "job") and cause other test failures.
+	 */
+	dt = k_uptime_get_32() - start_ms;
+	zassert_true(dt < 3000,
+		     "1cpu test took too long (%d ms)", dt);
+	arch_irq_unlock(key);
+}
+
+void z_impl_z_test_1cpu_start(void)
+{
+	cpuhold_active = 1;
+#ifdef CONFIG_THREAD_NAME
+	char tname[CONFIG_THREAD_MAX_NAME_LEN];
+#endif
+	k_sem_init(&cpuhold_sem, 0, 999);
+
+	/* Spawn N-1 threads to "hold" the other CPUs, waiting for
+	 * each to signal us that it's locked and spinning.
+	 *
+	 * Note that NUM_CPUHOLD can be a value that causes coverity
+	 * to flag the following loop as DEADCODE so suppress the warning.
+	 */
+	/* coverity[DEADCODE] */
+	for (int i = 0; i < NUM_CPUHOLD; i++)  {
+		k_thread_create(&cpuhold_threads[i],
+				cpuhold_stacks[i], CPUHOLD_STACK_SZ,
+				(k_thread_entry_t) cpu_hold, NULL, NULL, NULL,
+				K_HIGHEST_THREAD_PRIO, 0, K_NO_WAIT);
+#ifdef CONFIG_THREAD_NAME
```

Comment:
you can use `if (IS_ENABLED(CONFIG_THREAD_NAME)` and get rid of ifdef around tname variable declaration.

Reply:
Done